### ISSUE TITLE
Posting categories adding index to array

### ISSUE THREAD

**Issue by keithjgrant** [grantcodes/micropub#8]
When I post categories, the parameters have indexes added to the fieldname instead of empty brackets: `category[0]=foo&category[1]=bar`. It should be `category[]=foo&category[]=bar`

This is causing errors with my MP endpoint, which is reporting the category as "[Object object]"

**Comment by grantcodes** [comment]
Do you know what post method your using? I.e. json, form or multipart

I would guess form encoded? As that is the only one I wrote personally [/src/lib/object-to-form-data.js](https://github.com/grantcodes/micropub/blob/master/src/lib/object-to-form-data.js)

**Comment by Zegnat** [comment]
This looks like form encoded, and testing that script it does seem to give array items each their own numerical key. This is a bug, at least as far as Micropub is concerned.

Form encoding [per the URL Standard](https://url.spec.whatwg.org/commit-snapshots/5fc4c4c17f3077beda5e3647df73f4b1b0ba7084/#application/x-www-form-urlencoded) really only supports a straight list of tuples. The Micropub protocol [extends this with `[]`](https://www.w3.org/TR/2017/REC-micropub-20170523/#form-encoded-and-multipart-requests) as an indicator for “multi-valued properties”, but all other `[value]` strings are explicitly invalid:

> Note that the extent of the extensions to the `x-www-form-urlencoded` syntax is the addition of the square brackets to indicate an array. Syntax such as `foo[0]` and `foo[bar]` is not supported, and so clients are expected to use the [JSON syntax](https://www.w3.org/TR/2017/REC-micropub-20170523/#json-syntax) when posting more complex objects.

If you detect any nesting of objects in the object you are serialising, you should switch to creating a JSON request. If any properties contain an array, you append each array item to the FormData with the same name.

Maybe something like this is what should be done?

```javascript
function objectToFormData (object) {
  const formData = new FormData()
  Object.keys(object).forEach(key => {
    const data = object[key]
    if (Array.isArray(data)) {
      data.forEach(value => formData.append(key + '[]', value))
    } else if (data === Object(data)) {
      throw new TypeError('Object invalid as FormData for Micropub.')
    } else {
      formData.append(key, data)
    }
  })
  return formData
}
```

**Edit** to add that the above code is just for illustration. You probably want to check the type of values within the array. Instead my function just relies on `FormData.append` to force it all into a string regardless.

**Comment by grantcodes** [comment]
Should be fixed with the 1.3.8 release
